### PR TITLE
MQTT chat client and manager client

### PR DESCRIPTION
**Pull request by khanghh**
- added subscribe unsubscribe topic function and publish message function to UserClient
- added create & delete chat channel and ban user from chat channel function to AdminClient control panel

### CODE REVIEW

File: src/main/scala/uniq/mqttchat/client/AdminClient.scala [hunk 1]
```diff
@@ -0,0 +1,216 @@
+
+// copyright
+
+package uniq.mqttchat.client
+
+// scalastyle:off
```
Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/AdminClient.scala]: Change to `// scalastyle:off underscore.import`

Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/AdminClient.scala]: Merger all import with underscore `_` within `// scalastyle:off underscore.import` and `// scalastyle:on ...`

File: src/main/scala/uniq/mqttchat/client/AdminClient.scala [hunk 2]
```diff
@@ -0,0 +1,216 @@
+
+// copyright
```
Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/AdminClient.scala]: Change to `// Copyright (C) 2016-2017 Ark Maxim, Inc.`

File: src/main/scala/uniq/mqttchat/client/Message.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+final case class Message(
+  username: String,
+  content: String,
+  runtime: Long
+)
+
+object Message {
+  implicit val encoder: Encoder[Message] = deriveEncoder
+  implicit val decoder: Decoder[Message] = deriveDecoder
+}
```
Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/Message.scala]: Blank like after.

File: src/main/scala/uniq/mqttchat/client/UserClient.scala [hunk 4]
```diff
@@ -0,0 +1,111 @@
+
```
Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/UserClient.scala]: Remove this blank line.

File: src/main/scala/uniq/mqttchat/client/UserClient.scala [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  private def joinChanel: Unit = {
+    listAllChanel
+    val channel = StdIn.readLine("please select a chanel: ")
+    subscribe(channelTopic.format(channel))
+    doChat(channel)
+  }
+
+  private def doChat(channel: String): Unit = {
+    val msg = StdIn.readLine(s"$username: ")
+    if (msg != "exit"){
```
Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/UserClient.scala:69]: space before `{`.

File: src/main/scala/uniq/mqttchat/client/UserClient.scala [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+      doChat(channel)
+    }
+    else {
+      unsubscribe(channelTopic.format(channel))
+    }
+  }
+
+  private def listAllChanel: Unit = {
+    val result = httpRequest(userURL.format(username), GET)
+    val userRules = io.circe.parser.decode[MQTTUserACLRules](result.body).toTry.get
```
Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/UserClient.scala:85]: avoid using `get`. user `foldLeft` or `getOrElse` or simply `case match`.

File: src/main/scala/uniq/mqttchat/client/UserClient.scala [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            "%-30s\t%-5s\t%-5s\n",
+            acl.topic.replace(channelTopic.format(""), ""),
+            acl.read,
+            acl.write
+          )
+        }
+      }
+    )
+  }
+
```
Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/UserClient.scala]: Remove this blank line.

File: src/main/scala/uniq/mqttchat/client/UserClient.scala [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+object UserClient{
+  def apply(username: String, password: String): UserClient = new UserClient(username, password)
+}
```
Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/UserClient.scala]: Add blank line after.

File: src/main/scala/uniq/mqttchat/client/UserClient.scala [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            acl.write
+          )
+        }
+      }
+    )
+  }
+
+}
+
+object UserClient{
```
Comment (quybeans) [inline src/main/scala/uniq/mqttchat/client/UserClient.scala]: space before `{`.

File: src/main/scala/uniq/networking/mqtt/base/BaseClient.scala [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+import org.eclipse.paho.client.mqttv3.MqttCallback
+import org.eclipse.paho.client.mqttv3.MqttClient
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions
+import org.eclipse.paho.client.mqttv3.MqttMessage
+import uniq.mqttchat.client.Message
+import uniq.networking.mqtt.base.BaseClient._
+// scalastyle:on
+
+class BaseClient(username: String, password: String){
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/BaseClient.scala]: space before `{`.

File: src/main/scala/uniq/networking/mqtt/base/BaseClient.scala [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+  private[this] val settings = new MqttConnectOptions()
+
+  private[this] def setAuth(username: String, password: String): Unit ={
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/BaseClient.scala]: space before `{`.

File: src/main/scala/uniq/networking/mqtt/base/BaseClient.scala [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    if (username.length > 0 && password.length > 0)
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/BaseClient.scala]: follow this format for if else
```scala
if (...) {
  ...
} else {
  ...
}
```

File: src/main/scala/uniq/networking/mqtt/base/BaseClient.scala [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    connector.setCallback(callback)
+  }
+
+  def isConnected: Boolean = connector.isConnected
+
+  def printCommand: Unit = {
+    println("disconnect")
+    disconnect
+  }
+
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/BaseClient.scala]: Remove this blank line.

File: src/main/scala/uniq/networking/mqtt/base/BaseClient.scala [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  val userTopic = "uniq/mqtt/chat/client/%s"
+  val channelTopic = "uniq/mqtt/chat/chanel/%s"
+  val GET = "GET"
+  val DELETE = "DELETE"
+  val POST = "POST"
+  val areYouSure = "Are you sure [Y/N]? "
+  val Y = "Y"
+  val strChannel = "channel: "
+  val strUsername = "username: "
+
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/BaseClient.scala]: remove this blank line.

File: src/main/scala/uniq/networking/mqtt/base/BaseClient.scala [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+}
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/BaseClient.scala]: Always add a blank line at the end of file.

File: src/main/scala/uniq/networking/mqtt/base/model/MQTTACLRuleName.scala [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+import io.circe.Encoder
+import io.circe.generic.semiauto.deriveEncoder
+
+final case class MQTTACLRuleName(
+  username: String,
+  topic: String
+)
+
+object MQTTACLRuleName {
+  implicit val encoder: Encoder[MQTTACLRuleName] = deriveEncoder[MQTTACLRuleName]
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/model/MQTTACLRuleName.scala]: `implicit val encoder: Encoder[MQTTACLRuleName] = deriveEncoder` drop `[MQTTACLRuleName]` just `deriveEncoder` is enough.

File: src/main/scala/uniq/networking/mqtt/base/model/MQTTUser.scala [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+
+final case class MQTTUser(
+  username: String,
+  password: String
+)
+
+object MQTTUser {
+  implicit val encoder: Encoder[MQTTUser] = deriveEncoder[MQTTUser]
+  implicit val decoder: Decoder[MQTTUser] = deriveDecoder[MQTTUser]
+}
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/model/MQTTUser.scala]: Add blank line at the end of file.

File: src/main/scala/uniq/networking/mqtt/base/model/MQTTUser.scala [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+import io.circe.generic.semiauto.deriveEncoder
+import io.circe.generic.semiauto.deriveDecoder
+
+final case class MQTTUser(
+  username: String,
+  password: String
+)
+
+object MQTTUser {
+  implicit val encoder: Encoder[MQTTUser] = deriveEncoder[MQTTUser]
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/model/MQTTUser.scala]: drop `[MQTTUser]` at the end of this line, `= deriveEncoder` is enough.

File: src/main/scala/uniq/networking/mqtt/base/model/MQTTUser.scala [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+  implicit val decoder: Decoder[MQTTUser] = deriveDecoder[MQTTUser]
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/model/MQTTUser.scala]: same

File: src/main/scala/uniq/networking/mqtt/base/model/MQTTUserACLRules.scala [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+import io.circe.Decoder
+import io.circe.generic.semiauto.deriveDecoder
+
+final case class MQTTUserACLRules(
+  username: String,
+  password: String,
+  acls: List[MQTTUserACLRule]
+)
+
+object MQTTUserACLRules{
```
Comment (quybeans) [inline src/main/scala/uniq/networking/mqtt/base/model/MQTTUserACLRules.scala]: space before `{`.

**Review by quybeans — CHANGES_REQUESTED** [review]
image
Remove all `/target` files also.

File: src/main/scala/uniq/mqttchat/client/AdminClient.scala [hunk 21]
```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2016-2017 Ark Maxim, Inc.
+
+package uniq.mqttchat.client
+
+// scalastyle:off underscore.import
+import scala.io.StdIn
+
+import uniq.networking.mqtt.base.BaseClient
+import uniq.networking.mqtt.base.BaseClient._
```
Comment (chequochuu) [inline src/main/scala/uniq/mqttchat/client/AdminClient.scala:9]: wildcard import

File: src/main/scala/uniq/mqttchat/client/AdminClient.scala [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+
+  private def listAllUsers: Unit = {
+    getListAllUsers.foreach(x => println(x.username))
+  }
+
+  private def listAllACLRules: Unit = {
+    val result = httpRequest(aclURL, GET)
+    val listUser = io.circe.parser.decode[List[MQTTACLRule]](result.body).toTry.get
+    printf("%-10s\t%-30s\t%-5s\t%-5s\n", "username", "topic", "read", "write")
+    listUser.foreach(x =>
```
Comment (chequochuu) [inline src/main/scala/uniq/mqttchat/client/AdminClient.scala:71]: use more meaningful variable

File: src/main/scala/uniq/mqttchat/client/AdminClient.scala [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+    val username = StdIn.readLine(strUsername)
+    val result = httpRequest(userURL.format(username), GET)
+    if (result.code != 200) {
+      println(result.body)
+    }
+    else {
+      val userRules = io.circe.parser.decode[MQTTUserACLRules](result.body).toTry.get
+      println(s"ACL rules of $username")
+      printf("%-30s\t%-5s\t%-5s\n", "topic", "read", "write")
+      userRules.acls.foreach(x => printf("%-30s\t%-5s\t%-5s\n", x.topic, x.read, x.write))
```
Comment (chequochuu) [inline src/main/scala/uniq/mqttchat/client/AdminClient.scala:92]: x?

File: src/main/scala/uniq/mqttchat/client/AdminClient.scala [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+    val data = MQTTACLRule.encoder.apply(
+      MQTTACLRule(
+        username,
+        topic,
+        canRead,
+        canWrite
+      )
+    ).noSpaces
+    val result = httpRequest(aclURL, POST, data)
+    204 == result.code
```
Comment (chequochuu) [inline src/main/scala/uniq/mqttchat/client/AdminClient.scala]: Use a constant instead of number

**Comment by khanghh** [comment]
moved to PR#2.

**Pull request closed without merging**